runtime/src/main/cpp/Memory.cpp

```diff
@@ -616,38 +646,71 @@ inline uint32_t freeableSize(MemoryState* state) {
   } else if (UseCycleCollector) { // Possible root.
-    RuntimeAssert(!Atomic, "Cycle collector shalln't be used with shared objects yet");
-    RuntimeAssert(container->objectCount() == 1,
-        "cycle collector shall only work with single object containers");
+    RuntimeAssert(!Atomic && !container->shareable(), "Cycle collector shalln't be used with shared objects yet");
+    RuntimeAssert(container->objectCount() == 1, "cycle collector shall only work with single object containers");
     // We do not use cycle collector for frozen objects, as we already detected
     // possible cycles during freezing.
     // Also do not use cycle collector for provable acyclic objects.
     int color = container->color();
     if (color != CONTAINER_TAG_GC_PURPLE && color != CONTAINER_TAG_GC_GREEN) {
+      RuntimeAssert(!container->shareable(), "Shareable cannot be cycle collected");
       container->setColorAssertIfGreen(CONTAINER_TAG_GC_PURPLE);
       if (!container->buffered()) {
         container->setBuffered();
         auto state = memoryState;
         state->toFree->push_back(container);
-        if (state->gcSuspendCount == 0 && freeableSize(state) >= state->gcThreshold) {
+        GC_LOG("toFree is now %d\n", state->toFree->size());
+        if (state->gcSuspendCount == 0 &&
+            (state->toFree->size() >= state->gcThreshold || state->toRelease->size() >= state->gcThreshold)) {
           GarbageCollect();
         }
       }
     }
   }
 }
 
+inline void DecrementRC(ContainerHeader* container) {
+  bool useCycleCollector = container->tag() == CONTAINER_TAG_NORMAL;
+  if (container->decRefCount() == 0) {
+    FreeContainer(container);
+  } else if (useCycleCollector) {
+      RuntimeAssert(!container->shareable(), "Cycle collector shalln't be used with shared objects yet");
+      RuntimeAssert(container->objectCount() == 1, "cycle collector shall only work with single object containers");
+      // We do not use cycle collector for frozen objects, as we already detected
+      // possible cycles during freezing.
+      // Also do not use cycle collector for provable acyclic objects.
+      int color = container->color();
+      if (color != CONTAINER_TAG_GC_PURPLE && color != CONTAINER_TAG_GC_GREEN) {
+        container->setColorAssertIfGreen(CONTAINER_TAG_GC_PURPLE);
+        if (!container->buffered()) {
+          container->setBuffered();
+          memoryState->toFree->push_back(container);
+        }
+      }
+  }
+}
+
+template <bool CanCollect>
+inline void EnqueueDecrementRC(ContainerHeader* container) {
+  auto state = memoryState;
+  if (CanCollect) {
+    if (state->toRelease->size() > kMaxReleaseSetSize - 64 && state->gcSuspendCount == 0) {
+      GarbageCollect();
+    }
+  }
+  state->toRelease->push_back(container);
+}
+
 inline void initThreshold(MemoryState* state, uint32_t gcThreshold) {
   state->gcThreshold = gcThreshold;
-  state->toFree->reserve(gcThreshold);
+  state->toRelease->reserve(gcThreshold);
```

Comment:
Why was `toFree->reserve(gcThreshold)` removed?

Reply:
Do you suppose that it is unlikely to get filled up to threshold because `toRelease` will become full earlier?

Reply:
Yes, it's no longer even close to that threshold.